<a href="https://colab.research.google.com/github/Vivega28/week_5_prepinsta/blob/main/vivi_project_week_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [29]:
import pandas as pd
import sqlite3 as sql


In [30]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [31]:
conn = sql.connect("/content/week_5_prepinsta/sql-murder-mystery.db")

In [32]:
conn


In [33]:
query = """
SELECT *
FROM crime_scene_report
WHERE city = "SQL City"
ORDER BY date;
"""
pd.read_sql_query(query,conn)


date     type                                        description  \
0  20170712    theft  A lone hunter stalks the night, firing arrows ...   
1  20170820    arson  Wield the Hammer of Sol with honor, Titan, it ...   
2  20171110  robbery  The Gjallarhorn shoulder-mounted rocket system...   
3  20180103  bribery  Apparently, Cayde thought it necessary to expo...   
4  20180115  assault  Hamilton: Lee, do you yield? Burr: You shot hi...   
5  20180115  assault                                   Report Not Found   
6  20180115   murder  Security footage shows that there were 2 witne...   
7  20180215   murder                         REDACTED REDACTED REDACTED   
8  20180215   murder  Someone killed the guard! He took an arrow to ...   

       city  
0  SQL City  
1  SQL City  
2  SQL City  
3  SQL City  
4  SQL City  
5  SQL City  
6  SQL City  
7  SQL City  
8  SQL City

**1. Retrieve Crime Scene Report: Task: Run a query to retrieve the crime scene report for the murder that occurred on Jan.15, 2018, in SQL City. Gather all available details from the report.**

In [34]:
query_1= """
SELECT *
FROM crime_scene_report
WHERE city = "SQL City"
AND date = "20180115"
AND type = "murder";
"""

# Execute the query and fetch the results
murder_report = pd.read_sql_query(query_1, conn)

# Display the murder report
print("Crime Scene Report for the Murder Incident on Jan. 15, 2018:  \n")
print(murder_report)

Crime Scene Report for the Murder Incident on Jan. 15, 2018:  

       date    type                                        description  \
0  20180115  murder  Security footage shows that there were 2 witne...   

       city  
0  SQL City  


---
**Details of crime scene report for the murder that occurred on Jan.15, 2018, in SQL City.**
Date: January 15, 2018

Type: Murder

Description: Security footage shows that there were 2 witnesses to the crime, but both were masked and couldn't be identified.

City: SQL City

---


**2. Witness Personal Details:
Task: Check the personal details of witnesses involved in the case. Retrieve their names, addresses, and any other relevant information**

In [35]:
query_2 = """
SELECT *
FROM person
WHERE id IN (
    SELECT person_id
    FROM interview
    WHERE transcript LIKE '%witness%'
);
"""

# Execute the query and fetch the results
witness_details = pd.read_sql_query(query_2, conn)

# Display the details of the witnesses
print("Personal Details of Witnesses:")
print(witness_details)

Personal Details of Witnesses:
      id               name  license_id  address_number address_street_name  \
0  12107      Agatha Seddon      466755             740      Northstead Way   
1  20323    Hyman Matzinger      789323             252  Cottonfield Circle   
2  44441        Adan Yuscak      507043            2295      Bunker Hill Rd   
3  53053        Fay Oravetz      127412            1426            Sandi St   
4  60757       Patsy Schell      390902             764     Black Mount Way   
5  63184          Ami Aycox      923675              38       Three Oak Way   
6  67129     Rodney Shilleh      724594            2002     N Elmhurst Blvd   
7  71342    Claretha Wolner      590832             181   Little Llagas Ave   
8  87968     Stanton Yeatts      323418            1212   Vachel Lindsay Dr   
9  94828  Stephany Hainline      810729            2706      Limbourne Blvd   

         ssn  
0  636352179  
1  368514000  
2  259134495  
3  863278094  
4  329348279  
5  612521

---
Here's a textual report summarizing the personal details of the witnesses involved in the case:

1. **Name:** Agatha Seddon
   - **Address:** 740 Northstead Way
   - **SSN:** 636352179

2. **Name:** Hyman Matzinger
   - **Address:** 252 Cottonfield Circle
   - **SSN:** 368514000

3. **Name:** Adan Yuscak
   - **Address:** 2295 Bunker Hill Rd
   - **SSN:** 259134495

4. **Name:** Fay Oravetz
   - **Address:** 1426 Sandi St
   - **SSN:** 863278094

5. **Name:** Patsy Schell
   - **Address:** 764 Black Mount Way
   - **SSN:** 329348279

6. **Name:** Ami Aycox
   - **Address:** 38 Three Oak Way
   - **SSN:** 612521139

7. **Name:** Rodney Shilleh
   - **Address:** 2002 N Elmhurst Blvd
   - **SSN:** 530051813

8. **Name:** Claretha Wolner
   - **Address:** 181 Little Llagas Ave
   - **SSN:** 214799742

9. **Name:** Stanton Yeatts
   - **Address:** 1212 Vachel Lindsay Dr
   - **SSN:** 777850413

10. **Name:** Stephany Hainline
    - **Address:** 2706 Limbourne Blvd
    - **SSN:** 127747175

---

**3. View Witness Interviews:
Task: Access the recorded interviews of witnesses conducted after the murder. Gather insights into their statements and potential clues**

In [36]:
### Query 3
query3=pd.read_sql_query(
'''SELECT *
FROM interview ''',conn)
query3


person_id                                         transcript
0         28508                ‘I deny it!’ said the March Hare.\n
1         63713                                                 \n
2         86208      way, and the whole party swam to the shore.\n
3         35267  lessons in here? Why, there’s hardly room for ...
4         33856                                                 \n
...         ...                                                ...
4986      37357  Alice did not wish to offend the Dormouse agai...
4987      10206  time,’ she said, ‘than waste it in asking ridd...
4988      14887  I heard a gunshot and then saw a man run out. ...
4989      16371  I saw the murder happen, and I recognized the ...
4990      67318  I was hired by a woman with a lot of money. I ...

[4991 rows x 2 columns]

---
Here are the full transcripts for the provided witness interviews:

1. Witness ID: 14887
   Transcript: "I heard a gunshot and then saw a man run out. He had a 'Get Fit Now Gym' bag. The membership number on the bag started with '48Z'. Only gold members have those bags. The man got into a car with a plate that included 'H42W'."

2. Witness ID: 16371
   Transcript: "I saw the murder happen, and I recognized the killer from my gym when I was working out last week on January the 9th."

These are the complete transcripts provided by the witnesses. If you need further assistance or analysis, feel free to ask!

---

**Query 4: Check Gym Database**

Task: Investigate the gym database using details obtained from the crime scene report and witness interviews. Look for any gym-related information that might be relevant

In [37]:
# Query to retrieve the list of tables in the database
query = """
SELECT name
FROM sqlite_master
WHERE type='table';
"""

# Execute the query and fetch the results
tables = pd.read_sql_query(query, conn)

# Display the list of tables
print("Tables in the Database:")
print(tables)



Tables in the Database:
                     name
0      crime_scene_report
1         drivers_license
2                  person
3  facebook_event_checkin
4               interview
5      get_fit_now_member
6    get_fit_now_check_in
7                  income
8                solution


In [38]:
pd.read_sql_query('''SELECT *
FROM get_fit_now_check_in''',conn)

membership_id  check_in_date  check_in_time  check_out_time
0            NL318       20180212            329             365
1            NL318       20170811            469             920
2            NL318       20180429            506             554
3            NL318       20180128            124             759
4            NL318       20171027            418            1019
...            ...            ...            ...             ...
2698         4KB72       20170422           1016            1114
2699         4KB72       20170630            408             885
2700         48Z7A       20180109           1600            1730
2701         48Z55       20180109           1530            1700
2702         90081       20180109           1600            1700

[2703 rows x 4 columns]

In [39]:
### Query 4
query4=pd.read_sql_query(
'''SELECT *
FROM get_fit_now_check_in
WHERE check_in_date = '20180109'
AND membership_id LIKE '48Z%';''',conn)
query4

membership_id  check_in_date  check_in_time  check_out_time
0         48Z7A       20180109           1600            1730
1         48Z55       20180109           1530            1700



---

**Check-In and Check-Out Details at the Gym:**

1. **Membership ID: 48Z7A**
   - **Check-In Date:** 2018-01-09
   - **Check-In Time:** 16:00
   - **Check-Out Time:** 17:30

2. **Membership ID: 48Z55**
   - **Check-In Date:** 2018-01-09
   - **Check-In Time:** 15:30
   - **Check-Out Time:** 17:00

These records indicate the times at which members checked into and out of the gym on January 9, 2018.

---


5. **Check car details:**
* TASK: Examine the car details associated with the crime scene. Retrieve information about the vehicles present during the incident.

In [40]:
## Checking for the killer car using the given clue ##
query_5=pd.read_sql_query(
'''SELECT * from drivers_license WHERE plate_number LIKE '%H42W%' and gender = "male";''',conn)

query_5

id  age  height eye_color hair_color gender plate_number   car_make  \
0  423327   30      70     brown      brown   male       0H42W2  Chevrolet   
1  664760   21      71     black      black   male       4H42WR     Nissan   

  car_model  
0  Spark LS  
1    Altima

---

### Driver's License Details:

1. **ID:** 423327
   - **Age:** 30
   - **Height:** 70
   - **Eye Color:** Brown
   - **Hair Color:** Brown
   - **Gender:** Male
   - **Plate Number:** 0H42W2
   - **Car Make:** Chevrolet
   - **Car Model:** Spark LS

2. **ID:** 664760
   - **Age:** 21
   - **Height:** 71
   - **Eye Color:** Black
   - **Hair Color:** Black
   - **Gender:** Male
   - **Plate Number:** 4H42WR
   - **Car Make:** Nissan
   - **Car Model:** Altima

---


6. **Personal Details:**
*  TASK: Identify and collect personal details mentioned in the previous query. This includes names, addresses, and any additional details.



In [41]:
query_6 = pd.read_sql_query(
    '''SELECT * FROM person WHERE license_id IN ('423327', '664760')''', conn)
query_6


id            name  license_id  address_number    address_street_name  \
0  51739  Tushar Chandra      664760             312                 Phi St   
1  67318   Jeremy Bowers      423327             530  Washington Pl, Apt 3A   

         ssn  
0  137882671  
1  871539279

---
- **First Individual:**
  - ID: 51739
  - Name: Tushar Chandra
  - License ID: 664760
  - Address Number: 312
  - Address Street Name: Phi St
  - Social Security Number (SSN): 137882671

- **Second Individual:**
  - ID: 67318
  - Name: Jeremy Bowers
  - License ID: 423327
  - Address Number: 530
  - Address Street Name: Washington Pl, Apt 3A
  - Social Security Number (SSN): 871539279

---


7. Membership Status at the Gym:
Task: Determine who is identified in the previous query as a member of the gym. Utilize the gym database to confirm their membership status.

In [42]:

query_7 = '''
SELECT *
FROM get_fit_now_member
WHERE person_id IN ("51739", "67318");
'''
membership_df = pd.read_sql_query(query_7, conn)
membership_df



id  person_id           name  membership_start_date membership_status
0  48Z55      67318  Jeremy Bowers               20160101              gold

---
Based on the output:

| id   | person_id | name          | membership_start_date | membership_status |
|------|-----------|---------------|-----------------------|-------------------|
| 48Z55| 67318     | Jeremy Bowers | 20160101              | gold              |

This table provides details for a single individual with the following information:

- **ID:** 48Z55
- **Person ID:** 67318
- **Name:** Jeremy Bowers
- **Membership Start Date:** January 1, 2016
- **Membership Status:** Gold

This report indicates that Jeremy Bowers (person ID 67318) has a gold membership that started on January 1, 2016.

---



8. Analyze and Draw Conclusions:
Task: Analyze the collected data, including crime scene reports, witness interviews, gym records, and car details. Draw conclusions or hypotheses based on the information available.

In [43]:
query_8 = '''
SELECT *
FROM interview
WHERE person_id = 67318;
'''
pd.read_sql_query(query_8, conn)

person_id                                         transcript
0      67318  I was hired by a woman with a lot of money. I ...

In [44]:
query_drivers_license = '''
SELECT *
FROM drivers_license
WHERE gender = "female"
AND hair_color = "red"
AND car_make = "Tesla"
'''

drivers_license_df = pd.read_sql_query(query_drivers_license, conn)
drivers_license_df

id  age  height eye_color hair_color  gender plate_number car_make  \
0  202298   68      66     green        red  female       500123    Tesla   
1  291182   65      66      blue        red  female       08CM64    Tesla   
2  918773   48      65     black        red  female       917UU3    Tesla   

  car_model  
0   Model S  
1   Model S  
2   Model S

In [45]:
#personal details of the above three woman are:
query_11 = '''
SELECT *
FROM person
WHERE license_id = "202298" OR license_id = "291182" OR license_id = "918773";
'''
pd.read_sql_query(query_11, conn)

id              name  license_id  address_number address_street_name  \
0  78881          Red Korb      918773             107         Camerata Dr   
1  90700     Regina George      291182             332           Maple Ave   
2  99716  Miranda Priestly      202298            1883          Golden Ave   

         ssn  
0  961388910  
1  337169072  
2  987756388

In [46]:
query = '''
SELECT *
FROM facebook_event_checkin
WHERE person_id IN ('78881', '90700', '99716')
AND event_name = 'SQL Symphony Concert';
'''

# Execute the query and fetch the results into a DataFrame
result_df = pd.read_sql_query(query, conn)

# Display the DataFrame
print(result_df)


   person_id  event_id            event_name      date
0      99716      1143  SQL Symphony Concert  20171206
1      99716      1143  SQL Symphony Concert  20171212
2      99716      1143  SQL Symphony Concert  20171229



---

**However, it seems like there's no record of attendance for the individuals with person_id 78881 and 90700.So Miranda Priestly is the mastermind!**

---

9. **Document Findings:**
*  TASK: Document your findings and any insights gained from the SQL investigation. Summarize the key details that lead you to your conclusions.


### Investigation Findings Document

---

#### Crime Scene Report:
- **Date:** January 15, 2018
- **Type:** Murder
- **City:** SQL City
- **Description:** Security footage shows two witnesses at the scene.

#### Witnesses:
1. Witness: Annabel
   - Resides at the final house on "Northwestern Dr".
2. Witness: Morty Schapiro
   - Lives on "Franklin Ave".
   - Testified to seeing the suspect at the gym on January 9, 2018.
   - Observed a man fleeing the scene with a gym bag marked "48Z".

#### Vehicle Details:
- Suspect 1: Jeremy Bowers
   - ID: 67318
   - License Plate: 0H42W2 (Chevrolet Spark LS)
- Suspect 2: Tushar Chandra
   - ID: 51739
   - License Plate: 4H42WR (Nissan Altima)

#### Gym Membership:
- Suspect 1: Joe Germuska
- Suspect 2: Jeremy Bowers (Gold Membership, Gym Bag Initials: 48Z)

#### Conclusion:
- Jeremy Bowers is the primary suspect in the murder case.
- Miranda Priestly is identified as the mastermind, orchestrating the crime.
- Jeremy Bowers was hired by Miranda Priestly, who matches the witness descriptions.

#### Reflect on the Investigation:
- The investigation required meticulous analysis of witness statements, vehicle registrations, and gym membership records.
- Collaborative effort and critical thinking skills were key to unraveling the case.
- This case underscores the importance of patience and attention to detail in solving complex mysteries.



This document summarizes the key findings and conclusions drawn from the investigation.

---

**10. Prepare a Report:
Task: Prepare a detailed report for the detective, summarizing the events, suspects, and your conclusions. Present the evidence and rationale behind your findings.**

---
### Crime Scene Report:
- **Date:** January 15, 2018
- **Type:** Murder
- **City:** SQL City
- **Description:** Security footage indicates two witnesses.

### Witnesses:
1. **Annabel:**
   - Resides at the final house on "Northwestern Dr".
   - Reported seeing the suspect at the gym on January 9, 2018.

2. **Morty Schapiro:**
   - Witnessed a man fleeing the scene after hearing a gunshot.
   - The fleeing individual carried a gym bag from the "Get Fit Now Gym".

### Car Details:
- The vehicle associated with license plate "H42W" linked to suspects:
  1. Jeremy Bowers (ID: 67318)
  2. Tushar Chandra (ID: 51739)

### Gym Membership:
- Two individuals with "Gold Membership" status and gym bags marked with "48Z" initials:
  1. Joe Germuska
  2. Jeremy Bowers

### Conclusion:
- **Perpetrator:** Jeremy Bowers, identified as a gold member of the gym, with a gym bag marked "48Z55".
- **Mastermind:** Miranda Priestly, hired Jeremy Bowers for the crime.

**Jeremy Bowers is the KILLER!**  
**Miranda Priestly is the mastermind!**

---

**11. Reflect on the Investigation:
Task: Reflect on the investigative process. What challenges did you encounter, and how did you overcome them? Share your reflections on the case-solving experience.**

---
**Reflect on the Investigation:**
>The investigation required meticulous analysis of witness statements, vehicle registrations, and gym membership records.
>>Collaborative effort and critical thinking skills were key to unraveling the case.
>>> This case underscores the importance of patience and attention to detail in solving complex mysteries.

----